### DISCUSSION TITLE
Built-in 3D mesh trails

### DISCUSSION THREAD

**Discussion by KC098** [godotengine/godot-proposals#4020] [3D]
There's a lot of addons that accomplish this, but often not optimized or doesn't have the functionality you hope for. 
So why not integrate it? It has become something very useful lately that not a lot of people give too much thought on.

**Comment by Calinou** [comment]
Godot 4.0alpha already supports particle trails with GPUParticles (in both 2D and 3D). These rely on the RibbonTrailMesh and TubeTrailMesh primitive meshes, which have their own skeleton and can also be manually skinned to create 3D curved meshes.

**Reply by KC098** [reply]
How does it compare to like Unity's 'Trail renderer' component? For the looks of it, sounds like it needs more setup to achieve the same effect.

**Reply by HeavyRail0451** [reply]
You can use local coordinates for Particles and control length of trail with initial velocity. It's close to unity Trail system.

**Reply by OmarRinaz** [reply]
i would say its far from close to unity...

**Reply by Calinou** [reply]
Numerous proposals on trails have been opened since this discussion was last active, such as https://github.com/godotengine/godot-proposals/issues/6151 and https://github.com/godotengine/godot-proposals/issues/7082.

**Comment by joviol** [comment]
I think the issue of the current trail particle is that we do not really have control over the endpoint of the trail. In Unity we move trail's node and the trail will follow. In Godot, as far as I could make out (I am very new to it) we have a static node and we eject the trail from that point in a specific direction with an optional curve.  I found a project in github that does use the GPUParticle3D to do exactly that. but it has it limitations and probably does not scale.  A line renderer would also be a welcome addition especially if fully done on the GPU.

**Reply by Calinou** [reply]
See https://github.com/godotengine/godot/pull/115040, which may implement this in the end.
